### ISSUE TITLE
Handling of lockfile on crash/restart

### ISSUE THREAD

**Issue by GWustenstein** [uhppoted/uhppoted-tunnel#7]
I've spent the period since the last release beating uhppoted-tunnel in some pretty terrible ways; the vendor's app in monitor mode (5s updates) constantly, there's a python script running as a daemon to pull status and events parsing them into its own db (5s updates), there's another python daemon running as a UDP CRUD service on the same port that monitors the vendor's app MS Access database and if it sees any changes in the group and username tables it pushes them out to the remote site db using the same secure tunnel service (just uses a 'magic' device serial number, opcodes, and special data hash.)

Topology:
Win10-udp---udp-RPi-tls--------tls-VPS-tls------tls-RPi-udp---udp-UHPPOTE
         db daemon                                                           db daemon
    udp  db client                                                           udp db server
                                                                                    udb db client

This entire mess has been working pretty well all this time with the new release, and a few random network hits, the system recovers after about 10minutes or so.  And I've been doing development of the local and remote db applications, which means I've been nuking the event tables and re-syncing thousands of events several times a day over the tunnel all while the app is up and running.

Today, however, with several power hits hitting pretty much every node (and sub-node) in every permutation imaginable, the remote tunnel node got lost in such a way that it left the lockfile behind.  Looking through syslog, there's a ton of lines, starting off after a reboot:

Nov  4 15:32:14 netmon systemd[1]: session-5.scope: Succeeded.
Nov  4 15:55:11 netmon rc.local[841]: fatal error: concurrent map read and map write
Nov  4 15:55:11 netmon rc.local[841]: goroutine 5329 [running]:
Nov  4 15:55:11 netmon rc.local[841]: runtime.throw({0x3c4576, 0x21})
Nov  4 15:55:11 netmon rc.local[841]: #011runtime/panic.go:992 +0x5c fp=0x2856ee0 sp=0x2856ecc pc=0x47bf8
Nov  4 15:55:11 netmon rc.local[841]: runtime.mapaccess2_fast32(0x378680, 0x2800480, 0x306ee)
Nov  4 15:55:11 netmon rc.local[841]: #011runtime/map_fast32.go:62 +0x194 fp=0x2856ef0 sp=0x2856ee0 pc=0x202ac
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/router.(*Router).get(0x625478, 0x306ee)
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/router/router.go:92 +0x3c fp=0x2856f50 sp=0x2856ef0 pc=0x2b74c8
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/router.(*Switch).Received(0x280e988, 0x306ee, {0x2a73006, 0x1c, 0x7fa}, 0x2a700f0)
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/router/router.go:62 +0x24 fp=0x2856f64 sp=0x2856f50 pc=0x2b7108
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/tunnel/tls.(*tlsClient).received(0x28a00a0, {0x2a73000, 0x22, 0x800}, 0x280e988, {0x4415b4, 0x29a2000})
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/tunnel/tls/tls_client.go:162 +0x128 fp=0x2856fcc sp=0x2856f64 pc=0x339c24
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/tunnel/tls.(*tlsClient).listen.func1()
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/tunnel/tls/tls_client.go:150 +0x54 fp=0x2856fec sp=0x2856fcc pc=0x339a88
Nov  4 15:55:11 netmon rc.local[841]: runtime.goexit()
Nov  4 15:55:11 netmon rc.local[841]: #011runtime/asm_arm.s:824 +0x4 fp=0x2856fec sp=0x2856fec pc=0x7a140
Nov  4 15:55:11 netmon rc.local[841]: created by github.com/uhppoted/uhppoted-tunnel/tunnel/tls.(*tlsClient).listen
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/tunnel/tls/tls_client.go:149 +0xfc
Nov  4 15:55:11 netmon rc.local[841]: goroutine 1 [chan receive, 77 minutes]:
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/commands.(*Run).run(0x625900, 0x28a80d8, {0x440ad8, 0x284e180}, 0x280e6b8, 0x289c140)
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/commands/run.go:259 +0x228
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/commands.(*Run).exec(0x625900, 0x28a80d8, {0x440ad8, 0x284e180}, 0x280e6b8)
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/commands/run_linux.go:66 +0x27c
Nov  4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/commands.(*Run).Execute.func1(0x28a80d8, {0x440ad8, 0x284e180}, 0x280e6b8)
Nov  4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/commands/run_linux.go:33 +0x3c
.
then ends up
.
Nov  4 15:55:12 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/tunnel/tunnel.go:60 +0x24c
Nov  4 15:55:12 netmon rc.local[841]: goroutine 33 [select, 77 minutes]:
Nov  4 15:55:12 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/tunnel/tls.(*tlsClient).connect.func1()
Nov  4 15:55:12 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/tunnel/tls/tls_client.go:109 +0x154
Nov  4 15:55:12 netmon rc.local[841]: created by github.com/uhppoted/uhppoted-tunnel/tunnel/tls.(*tlsClient).connect
Nov  4 15:55:12 netmon systemd[1]: rc-local.service: Control process exited, code=exited, status=2/INVALIDARGUMENT

All of this seems to happen before the system has really booted.  Maybe the unit file needs to be modified to wait for full multi-user or maybe network?  On this RPi4 running Bullseye, I have about a 1min network wait to deal with the dumpster fire that is Debian-11/RPi's handling of the WiFi drivers.

In any event, there seems to be something missing in clean-up; if it crashes, it should take the lockfile with it, or the unit script should look for the pid and 86 both it and the file and start anew.

**Comment by uhppoted** [comment]
Those storms must have been quite something! 

There's quite a bit to unpack in that log:
- There might be a _mutex_ missing around one of the router tables (`concurrent map read and map write`) - will look into that
- The `goroutine 33 [select, 77 minutes] ... goexit` is something I've never seen - will have to dig into it to see exactly what is happening there
- Good point on waiting for _multi-user/network_ in the service file - will implement that (actually that needs to be implemented across the board)

The _lockfile_  needs some thinking about though - basically there's no way to guarantee that a crash won't leave a _lockfile_ behind so the recovery needs to happen on service start. There are a couple of options I can think of right off the bat:
- Allow the _lockfile_ to be opted out in the _conf_ file (e.g. `lockfile=(none)` or somesuch). Probably not such a good idea but it's simple
- Make the _lockfile_ a _soft lock_ a lá the fix for the _uhppoted-mqtt_ client _lockfile_. 
- Deleting it the service start (also a lá the _uhppoted-mqtt_ fix) . Am a bit reluctant to make that the default but it could be an 'advanced user' option
- An "_I'm alive_" hook so something like an external service or cron job could monitor the _daemon_ and clean up if necessary

Let me give it some thought..

**Comment by GWustenstein** [comment]
There is a lot more in the log, if you give me a few tidbits of keywords to look for I can try to ferret out more gotchas, but we'd all like the ultimate goal of this beast to be unattended and robust.  First and foremost, I need it to forward alarms and fortunately I'm using Pushover for that, which does a nice job of sending an event summary, a snap of the door, etc. to my phone/desktop/whatever so it does not depend on Tunnel.  But who's in the building, door control, siren silence, and all that rubbish is dependent on Tunnel and I don't want to have to futz with trying to manage a wayward process from my phone (or write a watchdog script if I can help it).

**Comment by uhppoted** [comment]
Thanks - think there's probably enough for me to start looking at it, but may ask you for more info once I get deeper into it. Or ... if there's nothing sensitive in the log file maybe attach it as a zip?

**Comment by uhppoted** [comment]
Hi,

Are you running _tunnel_ v0.8.2+ on the machine that crashed? I'm curious because:

- The `fatal error: concurrent map read and map write` issue was fixed in v0.8.2 (cf. https://github.com/uhppoted/uhppoted-tunnel/issues/4)
- [router.go:line 62](https://github.com/uhppoted/uhppoted-tunnel/blob/master/router/router.go#L61-L70) in v0.8.2+ doesn't seem to match the log:
```
Nov 4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/router.(*Switch).Received(0x280e988, 0x306ee, {0x2a73006, 0x1c, 0x7fa}, 0x2a700f0)
Nov 4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/router/router.go:62 +0x24 fp=0x2856f64 sp=0x2856f50 pc=0x2b7108
```

_(Iin the actual code `Received` is around router.go:67 and it hasn't changed since the v0.8.2 release)_

**Comment by GWustenstein** [comment]
That would be strange, because v0.8.1 does not run as a daemon.  Now you got me wondering if I managed some sort of frankeninstall, double-checking:

uhppoted-tunnel installed in /etc/uhppoted/uhpoted-tunnel and referenced as such in /etc/systemd/system/uhpoted-tunnel.service
MD5 of /etc/uhppoted/uhppoted-tunnel bcca16c461b9f2b721b0f3630ca74659

going over to the releases dir cloned from git as uhppoted_v0.8.2/uhppoted-tunnel, the MD5 is bcca16c461b9f2b721b0f3630ca74659, so I'm pretty sure it is what it is...

**Comment by GWustenstein** [comment]
No, I'm *much too lazy*, lol...
got from https://github.com/uhppoted/uhppoted/releases/download/v0.8.2/uhppoted_v0.8.2-arm7.tar.gz
just did a wget to a test directory, gunzip, tar xvf, to make sure it IS what I installed, and got the same MD5:
bcca16c461b9f2b721b0f3630ca74659

**Comment by uhppoted** [comment]
Ok, I get the same checksum now ... which is bizarre. I really hope it doesn't mean all those top level binaries are wrong - but you might want to switch to the _uhppoted-tunnel_ releases while I figure out what is going on.

**Comment by uhppoted** [comment]
FYI, turns out the checksums are different because the top-level release builds were not being built with `GO WORK=off` - not catastrophic but I'll fix the release script and add a checksum verification step.

Which, however, does not explain why the logs and the source code do not match..

**Comment by uhppoted** [comment]
Ok, at this point in time I am reasonably convinced that when your machine crashed it was running _tunnel v0.8.1_:

- the log entries match the source code of v0.8.1 exactly (and v0.8.2 not at all)
- I've dumped and checked the symbol tables for both the _uhppoted-tunnel v0.8.2_ and the _uhppoted v0.8.2_ releases and they're both correct for v0.8.2 (and not v0.8.1) so as far as I can tell neither of the executables would have produced those log entries.

I'm afraid I have absolutely no idea how that could have happened - but here are some things you could look at if you wanted to:

- The _tunnel_ daemon prints its version on startup - there may be something in the logs for 4th November like:
```
Nov 10 21:32:25 uhppoted uhppoted-tunnel[14728]: 2022/11/10 21:32:25 uhppoted-tunnel service v0.8.2 - Linux (PID 14728)
```
- `sudo systemctl status "uhppoted-tunnel"` should show the tunnel version when the daemon starts:
```
sudo systemctl status "uhppoted-tunnel-qwerty"
uhppoted-tunnel-qwerty.service - UHPPOTE UTO311-L0x access card controllers UDP tunnel service/daemon
     Loaded: loaded (/etc/systemd/system/uhppoted-tunnel-qwerty.service; disabled; preset: enabled)
     Active: active (running) since Thu 2022-11-10 21:32:25 UTC; 1min 48s ago
       Docs: https://github.com/uhppoted/uhppoted-tunnel
   Main PID: 14728 (uhppoted-tunnel)
      Tasks: 5 (limit: 1020)
     Memory: 1.3M
        CPU: 7ms
     CGroup: /system.slice/uhppoted-tunnel-qwerty.service
             └─14728 /opt/uhppoted/uhppoted_v0.8.2/uhppoted-tunnel --lockfile /var/uhppoted/uhppoted-tunnel-qwerty.pid --in udp/listen:0.0.0.0:60000 --out tcp/server:0.0.0.0:12345

Nov 10 21:32:25 uhppoted systemd[1]: Started UHPPOTE UTO311-L0x access card controllers UDP tunnel service/daemon.
Nov 10 21:32:25 uhppoted uhppoted-tunnel[14728]: 2022/11/10 21:32:25 uhppoted-tunnel service v0.8.2 - Linux (PID 14728)
```
- Is (was ?) the service file (_/etc/systemd/system/uhppoted-tunnel.service_) pointing to the correct executable?

I'm going to look at the `goroutine 33 [select, 77 minutes]:` thing just to try understand that - don't think there's much else I can offer around the _fatal error: concurrent map read and map_ write log entry.

**Comment by GWustenstein** [comment]
Going back in time to find when this run was launched before this epic storm...
Oct 24 13:32:23 netmon systemd[1]: Started UHPPOTE UTO311-L0x access card controllers UDP tunnel service/daemon.
Oct 24 13:32:23 netmon uhppoted-tunnel[5463]: 2022/10/24 13:32:23 uhppoted-tunnel service v0.8.2 - Linux (PID 5463)
Oct 24 13:34:06 netmon systemd[1]: session-606.scope: Succeeded.
(PID 5463 was up for nearly a week until the storm hit, then it was rock-n-roll for several hours, several power and network outages).

Went forward in the logs and pulled the part where it crashed when it found the stale pid file...
Nov  4 21:15:00 netmon systemd[1]: Started UHPPOTE UTO311-L0x access card controllers UDP tunnel service/daemon.
Nov  4 21:15:00 netmon uhppoted-tunnel[1563]: 2022/11/04 21:15:00 uhppoted-tunnel service v0.8.2 - Linux (PID 1563)
Nov  4 21:15:01 netmon uhppoted-tunnel[1563]: ERROR: PID lockfile '/var/uhppoted/uhppoted-tunnel.pid' already in use
Nov  4 21:15:01 netmon systemd[1]: uhppoted-tunnel.service: Main process exited, code=exited, status=1/FAILURE
Nov  4 21:15:01 netmon systemd[1]: uhppoted-tunnel.service: Failed with result 'exit-code'.
Nov  4 21:16:25 netmon systemd[1]: Started UHPPOTE UTO311-L0x access card controllers UDP tunnel service/daemon.
Nov  4 21:16:25 netmon uhppoted-tunnel[1591]: 2022/11/04 21:16:25 uhppoted-tunnel service v0.8.2 - Linux (PID 1591)

This is the RPi-4/Debian-11 (Bullseye) build, pulled as bin off git, as lazy as possible.
The pivot node is x86 running on a VPS, but still Deb-11, tunnel v0.8.2 if that matters at all, it's been rock-solid with no crashes. The x86/win10 side has been solid as well.  If I were a bettin' man, I'd say go/arm is a little effed...

**Comment by uhppoted** [comment]
Ok, am going to admit to being completely baffled by the _concurrent map access_ issue:

> Nov 4 15:55:11 netmon rc.local[841]: fatal error: concurrent map read and map write
> Nov 4 15:55:11 netmon rc.local[841]: goroutine 5329 [running]:
> Nov 4 15:55:11 netmon rc.local[841]: runtime.throw({0x3c4576, 0x21})
> Nov 4 15:55:11 netmon rc.local[841]: #011runtime/panic.go:992 +0x5c fp=0x2856ee0 sp=0x2856ecc pc=0x47bf8
> Nov 4 15:55:11 netmon rc.local[841]: runtime.mapaccess2_fast32(0x378680, 0x2800480, 0x306ee)
> Nov 4 15:55:11 netmon rc.local[841]: #011runtime/map_fast32.go:62 +0x194 fp=0x2856ef0 sp=0x2856ee0 pc=0x202ac
> Nov 4 15:55:11 netmon rc.local[841]: github.com/uhppoted/uhppoted-tunnel/router.(*Router).get(0x625478, 0x306ee)
> Nov 4 15:55:11 netmon rc.local[841]: #011github.com/uhppoted/uhppoted-tunnel/router/router.go:92 +0x3c fp=0x2856f50 sp=0x2856ef0 pc=0x2b74c8

As far as I can see that can only occur in v0.8.1. Am going to put that on the back burner for now (pending another storm and/or inspiration :-)) and focus on the _lockfile_ problem which is at least tangible.

**Comment by uhppoted** [comment]
### TODO

- [ ] Abstract Lockfile as _interface_ and move to _lib_
- [ ] Default to ephemeral _tmp_ folder for lockfiles
- [ ] Default to `flock` for Linux and MacOS
- [ ] _(optional)_ 'soft' lock
- [ ] _(optional)_ 'socket' lock
- [ ] Rework other modules to use _lib_ `Lockfile`
    - [ ] _uhppoted-rest_
    - [ ] _uhppoted-mqtt_
    - [ ] _uhppoted-httpd_
    - [ ] _uhppoted-app-wildapricot_
    - [ ] _uhppoted-app-s3_
    - [ ] _uhppoted-app-sheets_

### NOTES

1. https://github.com/golang/go/issues/33974
2. https://go.googlesource.com/proposal/+/master/design/33974-add-public-lockedfile-pkg.md
3. https://perl.plover.com/yak/flock
4. https://stackoverflow.com/questions/5210945/atomic-file-creation-on-linux      
5. https://www.diskodev.com/posts/linux-atomic-operations-on-files
6. https://stackoverflow.com/questions/29261648/atomic-writing-to-file-on-linux
7. https://stackoverflow.com/questions/34873151/how-can-i-delete-a-unix-domain-socket-file-when-i-exit-my-application
8. https://gavv.net/articles/unix-socket-reuse
9. https://stackoverflow.com/questions/7405932/how-to-know-whether-any-process-is-bound-to-a-unix-domain-socke
10. https://www.unix.com/man-page/centos/1/lockfile
11. https://security.stackexchange.com/questions/11976/can-unix-domain-sockets-be-locked-by-user-id
12. https://pythonhosted.org/lockfile/lockfile.html

**Comment by uhppoted** [comment]
That's quite an area you live in! Have you considered using something like e.g. Tailscale for the SSH?

re. lockfile. Am working on it - turns out there's a lot more to lock files than one would think (the official [Go implementation](https://github.com/golang/go/issues/33974) has been stuck in limbo for like 5 years). However - as an interim fix you could set the lock file to live in the _/tmp_ folder which is (supposedly) cleared on reboot e.g.
```
uhppoted-tunnel --lockfile /tmp/uhppoted-tunnel.lock .....
```

re. camera. Huh! Interesting idea!! And yeah, a Pi 4 is way overkill for most of this stuff :-).

**Comment by uhppoted** [comment]
Tailscale is neat and does seem to have pretty much got it right - unlike all the other hacky and marginal solutions that peope have resorted to over the years.

Although I see they do have a [CVE](https://emily.id.au/tailscale) today :-):

**Comment by GWustenstein** [comment]
Just read through that, since my install was today, I have the latest, but #7 is nasty, like log4j nasty.  Not running http through it, but ran the curl tests to be sure there are no gotchas.

While I'm at the task of modifying the system d unit file, is there a directive to have it shut off logging the packet transactions (or at least point them to /dev/null?)  Since this thing is running headless and mostly in RAM, I want to minimize SD use from logging and keep it only for alarm event recordings.  It's a high endurance card, but still logs tend to scribble blocks like crazy...

**Comment by uhppoted** [comment]
Ref. [#7](https://github.com/uhppoted/uhppoted-tunnel/issues/7). Eeesh, security researchers do **not** think in straight lines - looks like Tailscale is more or less on top of it though.

You managed to figure out logging to _/dev/null_ ? By default _tunnel_ logs to _stdout_ and _stderr_ which means the service typically logs to _syslog_ (depending on how your system is configured of course). Editing the command-line in the unit file to redirect _stdout_ to _/dev/null_ should do it.

**Comment by GWustenstein** [comment]
logging:  I saw the log level directive in the description, set it to warn in the Service description.

Hey, here's a weird inconsistency; In my typical laziness I pull the 0.8.2 arm linux binary to load up on the pi-zero-w.  It crashes out with Illegal Instruction, so I go, oh, built with 64bit.  No big deal, install golang and build it from source.  Pi-zero, so go make dinner, it'll be a *while*...  come back, stick the binary where I typically do (as /etc/uhppoted/uhppoted-tunnel) only this time when I go to deamonize it, it complains about not finding the .toml file.  That's funny, the 64bit binary downloaded from git does not do that on the pi4.  Made a uhppoted-tunnel.toml with some defaults I liked, including cert and key file locations, etc.  Well, it ignores those directives, but it stopped complaining.  

TL;DR: The .toml madness hints there is something out of sync with the pre-built binaries on git and the source.

**Comment by GWustenstein** [comment]
OMG, you're going to love this, and I feel like an idiot:  I have no idea how this happened, but there were both versions of uhppoted-tunnel running concurrently on this system, and looking at /etc/systemd/system I don't see how it is even possible.  It was hidden because I am running different client sessions to different hosts, so I expect to see a few out there when running a ps, but until I stopped all the services and saw one still hanging around, and pointing to /home/pi which is not even an active user directory(!) I got suspicious.  

This Pi is definitely haunted...  mentioned before this was a Buster->Bullseye upgrade, which is contraindicated due to hits on the networking stack, well, methinks there's other weirdness with apt and make.  This one gets wiped and a do-over.  The pi-zero is now up and running and is a clean build from 0.8.2.

**Comment by uhppoted** [comment]
Ha! I **knew** it :-) :-) :-) - but basically '_Whew!_' - am really, really happy to have that very puzzling anomaly explained.

Oh, BTW - figured out the hack for routing service logs to _/dev/null_ - if you run the service with the`--console` argument then the logs go to _stdout_ and you can redirect them to wherever you like. Not recommended though because if anything goes wrong there won't be a scrap of evidence to work with - `--log-level` is definitely the way to go.

**Comment by uhppoted** [comment]
Oh, that sounds great!!! Really glad to hear that, and thanks for the update!

The `flock` _syscall_ is the last piece of the puzzle I think - it's honestly the correct way to do locking and should in theory add a whole new level of robustness to the _lockfile_ thing .. been an interesting delve in the innards of systems.

**Comment by uhppoted** [comment]
v0.8.3 is available for whenever you have time to try it out (and a convenient storm). I'll leave this issue open for a bit..

**Comment by uhppoted** [comment]
'k, closing this - feel free to reopen if you need to.
